File: wstikkelorum/src/ast/literal/Variable.java
package ast.literal;

import ast.expression.Expression;
import ast.visitor.Visitor;

public class Variable extends Expression {
	private final String identifier;
	private final VariableType type;

	public Variable(int lineNumber, String identifier, VariableType type) {
		super(lineNumber);
		this.identifier = identifier;
		this.type = type;
	}

	public String getIdentifier() {
		return identifier;
	}

	public VariableType getType() {
		return type;
	}

	@Override
	public <T> T accept(Visitor<T> visitor) {
		return visitor.visit(this);
	}
}

File: JordyHeemskerk/QL/QL/Variable.swift
//
//  Variable.swift
//  QL
//
//  Created by Jordy Heemskerk on 24/02/16.
//  Copyright © 2016 Jordy Heemskerk. All rights reserved.
//

import Foundation
import SwiftParsec

struct Variable: Literal {
    
    let identifier: String
    let position: Position
    
    init(tupleInput: (String, Position)) {
        let (identifier, currentPosition) = tupleInput
        self.identifier = identifier
        self.position = currentPosition
    }
    
    func accept<T: ExpressionVisitor>(visitor: T) -> T.ExpressionReturnType {
        return visitor.visit(self)
    }
    
    func inferType(symbolTable: SymbolTable) -> Type? {
        return symbolTable.getVariableType(identifier)
    }
}